app/experimenter/nimbus-ui/src/components/FormBranches/reducer.tsx

```diff
@@ -0,0 +1,162 @@
+import { getExperiment_experimentBySlug } from "../../types/getExperiment";
+
+export function formBranchesReducer(
+  state: FormBranchesState,
+  action: FormBranchesAction,
+): FormBranchesState {
+  const newState = { ...state };
+  switch (action.type) {
+    case "addBranch":
+      return addBranch(newState);
+    case "removeBranch":
+      return removeBranch(newState, action);
+    case "updateBranch":
+      return updateBranch(newState, action);
+    case "removeFeatureConfig":
+      return removeFeatureConfig(newState);
+    case "setFeatureConfig":
+      return setFeatureConfig(newState, action);
+    case "setEqualRatio":
+      return setEqualRatio(newState, action);
+    default:
+      return newState;
+  }
+}
+
+export const REFERENCE_BRANCH_IDX = -1;
+
+export type FormBranchesState = {
+  equalRatio: boolean;
+  featureConfig: getExperiment_experimentBySlug["featureConfig"];
+  referenceBranch: getExperiment_experimentBySlug["referenceBranch"];
+  treatmentBranches: getExperiment_experimentBySlug["treatmentBranches"];
+};
```

Comment:
This could get slightly less verbose if you used `Pick<getExperiment_experimentBySlug, "featureConfig" | "referenceBranch" | "treatmentBranches"> & { equalRatio: boolean; };`, but this is perfectly fine (this is mostly me just excited to know about `Pick`)